### PR TITLE
[WIP] Computer Refactor

### PR DESCRIPTION
**Pull request by vache**
Will allow map-makers to include computers in their maps.  Currently only references existing hard-coded computer options with plans to include more in the future, that don't require digging into the code.

Also fixes a few issues, changes submap computers to an array instead of a single member.  This fixes a subtle, basically irrelevant, bug in the doctor's office, where two computers exist on the same submap.

### CODE REVIEW

File: src/computer.cpp [hunk 1]
```diff
@@ -47,13 +100,28 @@ computer &computer::operator=(const computer &rhs)
     security = rhs.security;
     name = rhs.name;
     mission_id = rhs.mission_id;
-    options = rhs.options;
-    failures = rhs.failures;
+    options.resize(rhs.options.size());
```
Comment (BevapDin) [inline src/computer.cpp]: Why are you not using the assignment operator of `std::vector`? Same in the copy constructor below.

Comment (vache) [inline src/computer.cpp]: I've been trying to debug an issue with options showing up in from-JSON buildings that lose all of their info somewhere in the process, and this was one of the attempts at tracking down that issue.  Also, it's still there so if you see anything glaringly wrong, let me know.

File: src/editmap.cpp [hunk 2]
```diff
@@ -1680,6 +1680,7 @@ int editmap::mapgen_preview( real_coords &tc, uimenu &gmenu )
                             memcpy( *destsm->frn, srcsm->frn, sizeof(srcsm->frn) ); // furniture
                             memcpy( *destsm->trp, srcsm->trp, sizeof(srcsm->trp) ); // traps
                             memcpy( *destsm->rad, srcsm->rad, sizeof(srcsm->rad) ); // radiation
+                            memcpy( *destsm->comp, srcsm->comp, sizeof(srcsm->comp) );
```
Comment (BevapDin) [inline src/editmap.cpp]: This is wrong, the computer struct contains complex objects like the string and vectors that must be copied or moved through their respective operators. Copying the bytes will not work, bad things will happen, I predict crashes. You should swap this like the cosmetics map below.

File: src/iexamine.cpp [hunk 3]
```diff
@@ -21,6 +21,30 @@ void iexamine::none(player *p, map *m, int examx, int examy)
...
+    if (p->has_trait("HYPEROPIC") && !p->is_wearing("glasses_reading")
+        && !p->is_wearing("glasses_bifocal") && !p->has_effect("contacts")) {
+        add_msg(m_info, _("You'll need to put on reading glasses before you can see the screen."));
+        return;
+    }
+
+    computer* c = m->computer_at(examx, examy);
+
+    if (c == NULL){
+        debugmsg("used computer at %d, %d, none found", examx, examy);
```
Comment (vache) [inline src/iexamine.cpp]: Pretend there's a return here.

Comment (KA101) [inline src/iexamine.cpp]: Sure, just tell the compiler that.  ;-)

File: src/computer.h [hunk 4]
```diff
@@ -67,6 +67,10 @@ enum computer_failure {
     NUM_COMPUTER_FAILURES
 };
 
+void initialize_computer_maps();
+static std::map<std::string, computer_action> string_to_action;
```
Comment (kevingranade) [inline src/computer.h]: You generally don't want to expose the actual container globally like this, you want to just expose accessors to it like computer::add_action(), computer::add_failure(), computer::get_action(), etc.

**Comment by Frost-wood** [comment]
I'm curious as to the status of the pr, as this is a very interesting pr, and this would allow us to move most of the mapgen to json.

**Comment by OzoneH3** [comment]
No progress for a long time, closing.

**Pull request closed without merging**